File: performance_test/test-storage-perf.py
import requests
import json
import time

def fetch_url_with_id(url, id):
    full_url = f"{url}/{id}"
    response = requests.get(full_url)
    if response.status_code == 200:
        return response.elapsed.total_seconds() * 1000
    else:
        return None

def fetch_all_urls(urls_and_ids, num_passes):
    for url_and_ids in urls_and_ids:
        url = url_and_ids["url"]
        ids = url_and_ids["ids"]
        print(f"URL: {url}")
        for id in ids:
            print(f"ID: {id}")
            for pass_num in range(num_passes):
                start_time = time.time()
                response_time = fetch_url_with_id(url, id)
                end_time = time.time()
                if response_time is not None:
                    print(f"Pass {pass_num+1} Response time (ms): {response_time:.2f}")
                else:
                    print(f"Pass {pass_num+1} failed")
                time.sleep(1)
            print("")
        print("")

if __name__ == "__main__":
    with open("input-py-nftsdcccc.json", "r") as f:
        urls_and_ids = json.load(f)
    num_passes = 5 # set the number of passes here
    fetch_all_urls(urls_and_ids, num_passes)

File: performance_test/test-storage-perf.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"time"
)

type URLAndIDs struct {
	URL string `json:"url"`
	IDs []int  `json:"ids"`
}

func fetchURLWithID(url string, id int) (float64, error) {
	fullURL := fmt.Sprintf("%s/%d", url, id)
	start := time.Now()
	resp, err := http.Get(fullURL)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusOK {
		elapsed := time.Since(start).Seconds() * 1000
		return elapsed, nil
	} else {
		return 0, fmt.Errorf("response status code: %d", resp.StatusCode)
	}
}

func fetchAllURLs(urlsAndIDs []URLAndIDs, numPasses int) {
	for _, urlAndIDs := range urlsAndIDs {
		url := urlAndIDs.URL
		ids := urlAndIDs.IDs
		fmt.Printf("URL: %s\n", url)
		for _, id := range ids {
			fmt.Printf("ID: %d\n", id)
			for passNum := 1; passNum <= numPasses; passNum++ {
				responseTime, err := fetchURLWithID(url, id)
				if err == nil {
					fmt.Printf("Pass %d response time (ms): %.2f\n", passNum, responseTime)
				} else {
					fmt.Printf("Pass %d failed: %s\n", passNum, err)
				}
				time.Sleep(1 * time.Second)
			}
			fmt.Println("")
		}
		fmt.Println("")
	}
}

func main() {
	jsonFile, err := os.Open("input-go-w3.json")
	if err != nil {
		panic(err)
	}
	defer jsonFile.Close()

	var urlsAndIDs []URLAndIDs
	jsonDecoder := json.NewDecoder(jsonFile)
	err = jsonDecoder.Decode(&urlsAndIDs)
	if err != nil {
		panic(err)
	}

	numPasses := 5 // set the number of passes here
	fetchAllURLs(urlsAndIDs, numPasses)
}
